pkg/grpc/api_crypto.go

```diff
@@ -0,0 +1,353 @@
+	// Process the request as a stream
+	return a.cryptoProcessStream(stream, reqProto, encOpts)
+}
+
+// DecryptAlpha1 decrypts a message using the Dapr encryption scheme and a key stored in the vault.
+func (a *api) DecryptAlpha1(stream runtimev1pb.Dapr_DecryptAlpha1Server) (err error) { //nolint:nosnakecase
+	// Get the first message from the caller containing the options
+	reqProto := &runtimev1pb.DecryptAlpha1Request{}
+	err = cryptoGetFirstChunk(stream, reqProto)
+	if err != nil {
+		a.Logger.Debug(err)
+		return err
+	}
+
+	// Validate required options
+	if reqProto.Options == nil {
+		err = messages.ErrBadRequest.WithFormat("first message does not contain the required options")
+		a.Logger.Debug(err)
+		return err
+	}
+
+	// Validate the request and get the component
+	component, err := a.CryptoValidateRequest(reqProto.Options.ComponentName)
+	if err != nil {
+		// Error is already logged
+		return err
+	}
+
+	// Options
+	encOpts := encv1.DecryptOptions{
+		UnwrapKeyFn: a.cryptoGetUnwrapKeyFn(stream.Context(), reqProto.Options.ComponentName, component),
+
+		// The next values are optional and could be empty
+		KeyName: reqProto.Options.KeyName,
+	}
+
+	// Process the request as a stream
+	return a.cryptoProcessStream(stream, reqProto, encOpts)
+}
+
+// Processes the request as a stream, encrypting or decrypting data.
+// For encryption, pass opts as an object of type encv1.EncryptOptions.
+// For decryption, pass opts as an object of type encv1.DecryptOptions.
+func (a *api) cryptoProcessStream(stream grpc.ServerStream, reqProto runtimev1pb.CryptoRequests, opts any) (err error) {
+	// Create a pipe to send the data to encrypt
+	inReader, inWriter := io.Pipe()
+
+	// Process the data coming from the stream
+	ctx := stream.Context()
+	go func() {
+		var (
+			readSeq   uint64
+			expectSeq uint64
+			payload   *commonv1pb.StreamPayload
+			readErr   error
+		)
+
+		// Process all chunks until EOF
+		for {
+			if ctx.Err() != nil {
+				inWriter.CloseWithError(ctx.Err())
+				return
+			}
+
+			// Get the payload from the chunk that was previously read
+			payload = reqProto.GetPayload()
+			if payload != nil {
+				readSeq, readErr = messaging.ReadChunk(payload, inWriter)
+				if readErr != nil {
+					inWriter.CloseWithError(readErr)
+					return
+				}
+
+				// Check if the sequence number is greater than the previous
+				if readSeq != expectSeq {
+					inWriter.CloseWithError(fmt.Errorf("invalid sequence number received: %d (expected: %d)", readSeq, expectSeq))
+					return
+				}
+
+				expectSeq++
+			}
+
+			// Read the next chunk
+			reqProto.Reset()
+			readErr = stream.RecvMsg(reqProto)
+			if errors.Is(readErr, io.EOF) {
+				// Receiving an io.EOF signifies that the client has stopped sending data over the pipe, so we can stop reading
+				inWriter.Close()
+				break
+			} else if readErr != nil {
+				inWriter.CloseWithError(fmt.Errorf("error receiving message: %w", readErr))
+				return
+			}
+
+			if reqProto.HasOptions() {
```

Comment:
add comment that this is executed only after the second chunk of data is read.

Reply:
The comment above says "Read next chunk"